### PR TITLE
cc-email-list: init component

### LINKED ISSUE #435: cc-email-list: init component
## Context

A component that displays the email addresses associated with a user account.

The component shows two things:

* the primary email address
* the secondary email addresses

### The primary email address

* The component displays a description of what this email address is used for.
* The component displays whether the primary email address has been verified. If not, the user has the ability to resend the confirmation email that will let him finish the verification process.

### The secondary email addresses

* The component displays a description of what these email addresses are used for.
* The user can add a new secondary email address.
* A secondary email address can be removed.
* A secondary email address can be marked as primary.

### "email" vs "email address"

In this component, we talk about two different things but the wording can be confusing. I'm talking about the email address versus the email sent for verification. We need to think about a good wording that will tackle ambiguity between these two notions. Both in the localized texts and in the properties and event namings.

Should we rename the whole component into `cc-email-address`?

### verification email

How do we call the email that will let the user finalize the process?
- verification email
- confirmation email
- validation email

How do we call the action for finalizing the process?
- verify email address
- validate email address

As for now, 
- API uses "confirmation_email".
- The email uses 
  - title: "Please validate you address… " (notice the typo :roll_eyes:)
  - body: "verify your email address" and "verify your address".
- The console uses "confirmation email", "validate this email", "resend account confirmation email".

### Notification messages

When the user clicks on the "resend confirmation email" button, we should notify him that the email has been sent and he needs to check his mailbox. Who is responsible for displaying this notification?

### Attributes/properties

| Property | Type   | Default value | Mandatory | Description |
|----------|--------|---------------|-----------|-------------|
| `model` | `EmailModel` | `null` | `false` | Model containing primary email address, secondary email addresses and state |
| `_addAddressInputValue` | `string` | `''` | `false` | Email address form input value |
| `_addAddressInputError` | `'empty'\|'invalid'\|'already-defined'\|'used'\|null` | `null` | `false` | Email address form input error key |

## Outputs

| Event | Data type | Description |
|-------|-----------|-------------|
| `send-confirmation-email` | `String` | Triggered when the user asks for sending a new confirmation email |
| `add` | `String` | Triggered when the user adds a new secondary email address |
| `delete` | `String` | Triggered when the user deletes a secondary email address |
| `mark-as-primary` | `String` | Triggered when the user marks a secondary email address as primary |

## Type Definitions

```typescript
export type PrimaryModelErrorType = "sendingConfirmationEmail";
export type PrimaryModelStateType = "sendingConfirmationEmail";

export type SecondaryModelErrorType = "adding";
export type SecondaryModelStateType = "adding";

export type SecondaryAddressErrorType = "markingAsPrimary" | "deleting";
export type SecondaryAddressStateType = "markingAsPrimary" | "deleting";

export interface EmailModel {
  primary: null | 'loadingError' | PrimaryModel;
  secondary: null | 'loadingError' | SecondaryModel;
}

export interface PrimaryModel {
  address: EmailAddress;
  state: PrimaryModelStateType;
  error: PrimaryModelErrorType;
}

export interface SecondaryModel {
  addresses: SecondaryAddressModel[];
  state: SecondaryModelStateType;
  error: SecondaryModelErrorType;
}

export interface SecondaryAddressModel {
  address: EmailAddress;
  state: SecondaryAddressStateType;
  error: SecondaryAddressErrorType;
}


export interface EmailAddress {
  value: string;
  verified: boolean;
}

```

## Data/APIs
* get the primary email address: `GET https://api.clever-cloud.com/v2/self`
* Send confirmation email: `GET https://api.clever-cloud.com/v2/self/confirmation_email`
* list secondary email addresses: `GET api.clever-cloud.com/v2/self/emails`
* add secondary email address: `PUT api.clever-cloud.com/v2/self/emails/{emailAddress}`
* delete secondary email address: `DELETE api.clever-cloud.com/v2/self/emails/{emailAddress}`
* mark secondary email address as primary: `PUT /v2/self/emails/<email>`

### Handling verified/unverified state

As for now, we cannot display the verified/unverified state because the API doesn't give us this information. In fact, we can only get verified email addresses from the API (unverified email addresses are never returned by the API).

We need the API to support:
- list all secondary email addresses (verified and unverified)
- each email address returned should have a `verified` boolean property.

### Resending confirmation email

As for now, the only way to resend a confirmation email is to submit the add form again. There is no explicit action for resending the confirmation email. And this is for the same reason as above: unverified email addresses are never returned by the API. Also, the API for such an action is missing.

See https://gitlab.corp.clever-cloud.com/core-team/cc-api/-/issues/743

## Stories & states

* Skeleton/loading
* Empty
* Loaded
* Loaded with a huge email addresses
* Waiting/sendingConfirmationEmail
* Waiting/adding
* Waiting/deleting
* Waiting/markingAsPrimary
* Error/loading
* Error/sendingConfirmationEmail
* Error/adding
* Error/deleting
* Simulation

## UX writing & i18n

* title: Secondary email addresses
* description: _TODO_
* delete button: just the icon without any label is ok.
* add button: Ajouter, Add
* error.add
* error.delete

## Wireframes and UI propositions

_TBD_

### LINKED ISSUE #628: cc-button: display waiting state when in link mode
When cc-button is in link mode, the waiting state is not materialized.

### PR DESCRIPTION
**Pull request by pdesoyres-cc**
Fixes #435

Fixes #628

Preview: https://clever-components-preview.cellar-c2.services.clever-cloud.com/cc-email/index.html?path=/story/%F0%9F%9B%A0-email-list-cc-email-list--default-story

- [x] remove mock API

### CODE REVIEW

**Comment by roberttran-cc** [comment]
Not much to say about the UI, look rather clean to me!

Some feedback:
- i'd always put the confirmation mail error message under the resend button;
- multiple component in the skeleton story needed?

And about the verification email wording, i vote for 'confirmation'! :+1:

**Comment by florian-sanders-cc** [comment]
Hey !

I think the UI looks good as it is :+1:.

A few details:

- we could try bringing the email icon closer to the email address, I feel like 1em is a bit too much, maybe 0.5em ? It's just a suggestion though, it's definitely up to you :+1: 
image
image
- "adding secondary" story, the email address field is empty and it would be a little more realistic if it contained a value (it's pretty insignificant though :joy:) 
- "simulations (secondary)" story, the error below the field should be displayed before setting components in waiting mode since we don't send the API call as long as the form contains an error.
- I find it disturbing that there are two block in stories, maybe a story for verified vs unverified ?

**Comment by hsablonniere** [comment]
## Notes following the 2022-09-14 meeting about state & smart

About the smart stuffs:

* console.warn when context is empty, I need to think about this and see if and when it's necessary
* I'm wondering if we really need the two separated functions `initComponent` and the other
  * It adds separation so maybe it's clearer but it also adds indirection
* I need to dig the `.ctrl.js` with the code but my first impression was "it seems too complex without much gain"
  * It's just a first impression :hugs:
  * I'm wondering if we're really going to write a `.ctrl.js` file for each component
  * I don't really understand how it simplify things or how it helps us to prevent errors
  * It's imported by the UI component and it relies on immer so technically, now all our components rely on immer
  * In a MVC or MV* world, I would argue that the controller is what we have in the smart definition, what you put in the `.ctrl.js` feels like both Model and Controller
  * I'm often in favor of WET, Write Everything Twice before diving into common/refactored helpers
    * It feels like the StateController, FormStateController and GenericStateController are premature optimization

About the state stuff:

* The work you've done on this is impressive :+1:
  * It's very close to what I want in the end, with the generics and common stuffs that impose consistency
* There's a bit too much nesting to my taste, especially when the names are too generic because of common/factorized types but I think it's close to our goal
* Many abstract/sub types mean the documentation from the outside could be very complex, we need to find the right balance

**Comment by roberttran-cc** [comment]
Well, overall agree with @hsablonniere last comment! :wink: 

Really impressive stuff: implementing things this thoroughly is so nice to make decisions, thanks so much. :bow:
On both parts (state and smart), imho and from a developer point of view, it brings a legitimate sense of safety but at the cost of being a little difficult to understand when discovering the code at first (and if I were to implement it).

:+1: on the need to find the right balance collectively!

**Comment by pdesoyres-cc** [comment]
PR is ready for review.

* Note1: it still contains the smart component in the md file so that you can test it easily.
* Note2: the APIs are mocked so that you won't be banned after several interactions.

Both are to be removed once the PR is approved.

File: src/components/cc-email-list/cc-email-list.stories.js
Comment (roberttran-cc) [inline src/components/cc-email-list/cc-email-list.stories.js:43]: Should we be more explicit with:
```suggestion
    {
      emails: {
        state: 'loading',
      },
    },
```

(and same in `simulationsWithPrimary` & `simulationsWithSecondary`)

File: src/components/cc-email-list/cc-email-list.stories.js
Comment (roberttran-cc) [inline src/components/cc-email-list/cc-email-list.stories.js:122]: If I understand correctly, `verified` property on secondary addresses are not needed... am I right? :thinking: 

(here and in other stories below)

File: src/translations/translations.fr.js
Comment (roberttran-cc) [inline src/translations/translations.fr.js:262]:
```suggestion
  'cc-email-list.secondary.address-input.error.used': `Cette adresse e-mail ne vous appartient pas`,
```

**Review by roberttran-cc — CHANGES_REQUESTED** [review]
Wow, we're reaaaaaaaally close to the end, that's so cool. :sunglasses: Some minor feedback nevertheless!

And a random idea (but I'd don't think we'll implement it): making email addresses clickable with `mailto:`. :eyes:

File: src/translations/translations.fr.js
Comment (roberttran-cc) [inline src/translations/translations.fr.js:240]:
```suggestion
  'cc-email-list.primary.action.resend-confirmation-email.error': ({ address }) => sanitize`Une erreur est survenue pendant l'envoi de l'e-mail de confirmation à l'adresse <strong>${address}</strong>.`,
```

**Review by roberttran-cc — CHANGES_REQUESTED** [review]

**Review by florian-sanders-cc — CHANGES_REQUESTED** [review]
:clap: Gg ! 
The component looks good, it's easy to use and it's well documented :+1: 

Just a few notes apart from the minor comments I've added:

# Email form with small viewport

I think we can improve how the form looks when the viewport is small:
image

We need to make the button wrap below the input field:
```css
form {
    align-items: start;
    display: flex;
    gap: 0 1em; // cancel the row gap
    flex-wrap: wrap; // wrap the button below
    justify-content: end; // align the button on the right like we do in other components
}

form > cc-input-text {
  flex: 1 1 19em; // make sure the input width is at least 19em (pick the value you think is appropriate)
}
```
image

If you think there is still too much space between the `cc-input-text` and `cc-button`, we can cancel its top margin with negative bottom margins on both these elements + bottom padding on the form element but I don't I don't like messing around with negative margins unless we really have to :roll_eyes:. 

I think we need to improve our solution for horizontal forms (`var(--cc-margin-top...`) because this is not the only case where this problem exists. I'll open an issue for this.

# Stories

It might be voluntary but there are no stories showing error messages for `already-defined` / `used` cases.
It would be easier to check the error message content if we had these but it's really not essential so I'll let you decide :+1:

File: src/components/cc-email-list/cc-email-list.js
Comment (florian-sanders-cc) [reply src/components/cc-email-list/cc-email-list.js:330]: Ah alright, I ran the a11y tests on `cc-orga-member` and realised that with `0.6` it passes the color contrast ratio criterion and with `0.5` it doesn't (probably why you went for `0.6` in the first place).

Considering this, should we switch to `0.6` everywhere ?

File: src/components/cc-email-list/cc-email-list.types.d.ts
Comment (Galimede) [inline src/components/cc-email-list/cc-email-list.types.d.ts:25]: There's a missing semicolon at the end.

**Review by Galimede — CHANGES_REQUESTED** [review]
Hiya, 
GG @pdesoyres-cc that's great and as @florian-sanders-cc that's well documented :clap: :heart_eyes: 
I've listed a few notes/typos below :wink:

File: src/translations/translations.en.js
Comment (roberttran-cc) [reply src/translations/translations.en.js:247]: Agree to sync on this! Good catch.

File: src/components/cc-email-list/cc-email-list.stories.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.stories.js:122]: You're right. It is an anticipation of the future.
We have only one `EmailAddress` interface for the primary address and the secondary addresses. I did not want to create two different interfaces (one with a `verified` property and another one without).
In order to respect this interface I have to specify a `verified` property even if not used.
I'll let this like that and very soon, the property will be used once we have the right API.

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:340]: Yes, we should consider switching to 0.6 everywhere... using a CSS var?

File: src/components/cc-email-list/cc-email-list.stories.js
Comment (roberttran-cc) [reply src/components/cc-email-list/cc-email-list.stories.js:126]: Nice, thanks for the explanation! :+1:

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.smart.js:116]: I will use `if (inputError != null)`

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:54]: I will use `email` everywhere

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:284]: good catch. thank you

File: src/components/cc-email-list/cc-email-list.js
Comment (roberttran-cc) [inline src/components/cc-email-list/cc-email-list.js:131]: Actually, we've got a little utility for this one!

```js
import { sortBy } from '../../lib/utils.js';
//...
return [...this.emails.value.secondaryAddresses].sort('address');
```

**Review by roberttran-cc — APPROVED** [review]

File: src/translations/translations.fr.js
Comment (roberttran-cc) [inline src/translations/translations.fr.js:258]:
```suggestion
  'cc-email-list.secondary.action.mark-as-primary.success': ({ address }) => sanitize`L'adresse e-mail <strong>${address}</strong> a bien été définie comme primaire.`,
```

**Review by roberttran-cc — CHANGES_REQUESTED** [review]

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (hsablonniere) [inline src/components/cc-email-list/cc-email-list.smart.js:22]: Interesting... I'm not sure about this yet but I like the let's try to improve all this verbosity.

I'd like the opinion of the team on this and if positive maybe discuss doing this elsewhere.

Note that the API calls that modify stuffs should not use the signal as we still want them to go through even if there's a navigation. I'll add notes below.

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (hsablonniere) [inline src/components/cc-email-list/cc-email-list.smart.js:71]: Can you use a body for the arrow function?

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (hsablonniere) [inline src/components/cc-email-list/cc-email-list.smart.js:240]: We don't need the signal on those as they are update calls and we want them to go through even if there's a navigation.

File: src/translations/translations.en.js
Comment (hsablonniere) [inline src/translations/translations.en.js:252]: :gb:  organisation with an S

**Review by hsablonniere — CHANGES_REQUESTED** [review]
Impressive work :clap: 

* I think I would have aligned the mark as primary and delete button with the email (and not the icon)
  * Same with the verified label on the primary email with long address
  * Maybe we already discussed this, it's something up to discussion, but both are OK with me ;-)
* I'm not sure about the simulations as we won't add a secondary address directly in the list without a validation first, right?

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (hsablonniere) [inline src/components/cc-email-list/cc-email-list.smart.js:206]: Sad but it seems the clever-client does not provide such a call.

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (hsablonniere) [inline src/components/cc-email-list/cc-email-list.smart.js:220]: This one is available on `user` as `todo_getConfirmationEmail`. Can you use it with an alias?

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:166]: With the `update` method, the focus behavior is entirely driven by the data model. When the error comes from the server (from the smart component) I don't need to focus the input with a public method explicitly. Instead, I just set the right error code in the data model, and the component focuses the text input automatically.

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:117]: you're right

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:234]: true, We don't really need it here

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:178]: much cleaner. Thank you

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:234]: I would say: If it's a big list with many refresh => use `repeat`. Otherwise, you can stick on `map`.
This being said, when using react you just don't have a choice: you must use a repeat-like function to display lists.

File: src/components/cc-email-list/cc-email-list.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.js:349]: In fact, I just want this for secondary addresses to avoid this:
image
By the way, if you have a better solution to handle that, I'd be glad to hear about it.

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.smart.js:22]: Some arguments for this:

* improve verbosity
* in the code, you manipulate a variable named `api` which guides the reader about what is involved
* If you want a temporary mocked `api` it may be easier to write.

File: src/translations/translations.en.js
Comment (pdesoyres-cc) [reply src/translations/translations.en.js:252]: There is another occurrence of `organization` on `cc-header-orga.error`. If it's ok for you, I'll fix it with this PR

File: src/components/cc-email-list/cc-email-list.smart.js
Comment (pdesoyres-cc) [reply src/components/cc-email-list/cc-email-list.smart.js:220]: thanks

**Comment by pdesoyres-cc** [comment]
First of all, I want to thank the team for the reviews. I really like our process: it makes my day-to-day work very satisfying :heart: 

About the email icon, several people don't like the `envelope` icon from remix icon (personally, I don't like it either). So I tried to modify the icon to make it more rectangular:

image
becomes
image

Please let me know if your ok with allowing some customization when we don't like the icon proposed by remix.

> I think I would have aligned the mark as primary and delete button with the email (and not the icon)
> * Same with the verified label on the primary email with long address
> * Maybe we already discussed this, it's something up to discussion, but both are OK with me ;-)

I can't remember if we talked about that. Here is a preview so that we can vote:
* :lion: [align with icon](https://clever-components-preview.cellar-c2.services.clever-cloud.com/cc-email/index.html?path=/story/%F0%9F%9B%A0-profile-cc-email-list--data-loaded-with-huge-email)
* :zebra: [align with email](https://clever-components-preview.cellar-c2.services.clever-cloud.com/cc-email-align/index.html?path=/story/%F0%9F%9B%A0-profile-cc-email-list--data-loaded-with-huge-email)

However, I don't understand your point about the alignment for the primary email. As of now, the verified label wraps under the address.

>    I'm not sure about the simulations as we won't add a secondary address directly in the list without a validation first, right?

Your right @hsablonniere, the simulations are useless: I removed them

**Review by florian-sanders-cc — CHANGES_REQUESTED** [review]
Small detail because I forgot to check the part about `cc-button` link style.
Apart from this, all good, Gg ! :clap:

File: src/components/cc-button/cc-button.js
Comment (pdesoyres-cc) [reply src/components/cc-button/cc-button.js:469]: Good catch!

**Comment by pdesoyres-cc** [comment]
Found a very weird bug on Safari and only when integrated into the Console. Ths skeleton animation doesn't stop immediately after the cc-badge 'skeleton' property is switched to false:
- the component loads: the badge is in skeleton mode
- the API is called: the badge is still in skeleton mode
- the API responds: the badge is no more in skeleton mode
  => :bug: the badge is well displayed BUT the skeleton animation is still played for a certain amount of time.

When the primary email is validated:
cc-email-list-safari-bug-1

When the primary email is not validated (here the email also still has the animation running):
cc-email-list-safari-bug-2

After struggling, I found that feeding the cc-badge slot with a TextNode is something Safari doesn't like. The workaround I found is as follows:
```
<cc-badge>Verified</cc-badge>
```
becomes
```
<cc-badge><span>Verified</span></cc-badge>
```

To track the reason for this workaround, I will add a comment in the code with a link to this PR comment.

**Pull request merged**
